File: Client/src/app/source/helpers/order-helper.ts
import { Expect } from "jest-editor-support";
import { EnumOrderStatus } from "../models/enums/order-status-enum";
import { EnumDeliveryMethod } from "../models/enums/delivery-method-enum";
import { OrderResponse } from "../models/dtos/entities/order-response";
import { LocalStorageService } from "../../services/common/local-storage.service";

export class OrderHelper {
    static nextStep(currentStatus: string, deliveryMethod: string) : string {

        if (currentStatus === EnumOrderStatus.Delivered || currentStatus === EnumOrderStatus.Canceled || currentStatus === EnumOrderStatus.Declined) {
            return currentStatus;
        }
        
        if (deliveryMethod == EnumDeliveryMethod.TakeAway) {
            switch(currentStatus) {
                case EnumOrderStatus.Started:
                    return EnumOrderStatus.Accepted;
                case EnumOrderStatus.Accepted:
                    return EnumOrderStatus.Cooking;
                case EnumOrderStatus.Cooking:
                    return EnumOrderStatus.Ready;
                case EnumOrderStatus.Ready:
                    return EnumOrderStatus.Delivered;
                default:
                    throw new Error("Estatus no válido");                     
            }
        }

        if (deliveryMethod == EnumDeliveryMethod.ForDelivery) {
            switch(currentStatus)
            {
                case EnumOrderStatus.Started:
                    return EnumOrderStatus.Accepted;
                case EnumOrderStatus.Accepted:
                    return EnumOrderStatus.Cooking;
                case EnumOrderStatus.Cooking:
                    return  EnumOrderStatus.Ready;
                case EnumOrderStatus.Ready:
                    return EnumOrderStatus.OnTheRoad;
                case EnumOrderStatus.OnTheRoad:
                    return EnumOrderStatus.Delivered;
                default:
                    throw new Error("Estatus no válido");
            }
        }

        throw new Error("Método de entrega/Estatus no válido");        
    }

    static canUserChangeStatus(order: OrderResponse, localStorageService: LocalStorageService) {
        return true;
    }
}

File: Server/Source/Helpers/OrderHelper.cs
using Server.Source.Exceptions;
using Server.Source.Extensions;
using Server.Source.Models.Enums;

namespace Server.Source.Helpers
{
    public static class OrderHelper
    {
        public static EnumOrderStatus NextStep(string currentStatus, string deliveryMethod)
        {
            EnumOrderStatus? enumStatus = null;

            var values = Enum.GetValues(typeof(EnumOrderStatus));
            foreach (EnumOrderStatus value in values)
            {
                if (value.GetDescription() == currentStatus)
                {
                    enumStatus = value;
                    break;
                }
            }

            if (enumStatus == EnumOrderStatus.Delivered)
            {
                throw new EatSomeInternalErrorException(EnumResponseError.OrderWasDelivered);
            }

            if (deliveryMethod == EnumDeliveryMethod.TakeAway.GetDescription())
            {
                switch (enumStatus)
                {
                    case EnumOrderStatus.Received:
                        return EnumOrderStatus.Accepted;
                    case EnumOrderStatus.Accepted:
                        return EnumOrderStatus.Cooking;
                    case EnumOrderStatus.Cooking:
                        return EnumOrderStatus.Ready;
                    case EnumOrderStatus.Ready:
                        return EnumOrderStatus.Delivered;
                    default:
                        throw new EatSomeInternalErrorException(EnumResponseError.InternalServerError);
                }
            }

            if (deliveryMethod == EnumDeliveryMethod.ForDelivery.GetDescription())
            {
                switch(enumStatus)
                {
                    case EnumOrderStatus.Received:
                        return EnumOrderStatus.Accepted;
                    case EnumOrderStatus.Accepted:
                        return EnumOrderStatus.Cooking;
                    case EnumOrderStatus.Cooking:
                        return  EnumOrderStatus.Ready;
                    case EnumOrderStatus.Ready:
                        return EnumOrderStatus.OnTheRoad;
                    case EnumOrderStatus.OnTheRoad:
                        return EnumOrderStatus.Delivered;
                    default:
                        throw new EatSomeInternalErrorException(EnumResponseError.InternalServerError);
                }
            }           

            throw new EatSomeInternalErrorException(EnumResponseError.InternalServerError);
        }

        public static EnumOrderStatus Canceled(string currentStatus)
        {
            return EnumOrderStatus.Canceled;
        }

        internal static EnumOrderStatus Declined(string currentStatus)
        {
            return EnumOrderStatus.Declined;
        }
    }
}
